Partial and copartial functions

EgbertRijke opened the pull request:
This PR adds partial and copartial functions to the library. Partial functions were needed to define partial sequences, which are ultimately needed for some OEIS entries. Copartial functions are simply the dual of partial functions, which I came up with while formalizing partial functions.

Note that partial functions are defined using the open modalities, so naturally copartial functions are defined using the closed modalities. They specify where a function is prohibited. The terminology in the files about copartial functions is not established in the literature (as far as I know), and should be considered experimental.

fredrik-bakke commented on src/foundation/partial-functions.lagda.md:
````diff
@@ -0,0 +1,71 @@
+words, a partial function is a function
+
+```text
+  A → Σ (P : Prop), (P → B).
````
Why not immediately consider partial dependent functions and all the other variants?

```text
  (x : A) → Σ (P : Prop), (P → B x).
```

If you don't need this then please don't go out of your way to add them just because I mentioned it.

EgbertRijke replied:
Ah, simply because I didn't think of them yet. Great idea!

fredrik-bakke commented on src/foundation/partial-functions.lagda.md:
````diff
@@ -0,0 +1,71 @@
+```agda
+partial-function :
+  {l1 l2 : Level} (l3 : Level) → UU l1 → UU l2 → UU (l1 ⊔ l2 ⊔ lsuc l3)
+partial-function l3 A B = A → partial-element l3 B
````
I.e. here `(x : A) → partial-element l3 (B x)`

fredrik-bakke commented on src/foundation/copartial-functions.lagda.md:
```diff
@@ -0,0 +1,72 @@
+
+## Idea
+
+A {{#concept "copartial function"}} from `A` to `B` is a map from `A` into the
```
```suggestion
A {{#concept "copartial function" Agda=copartial-function}} from `A` to `B` is a map from `A` into the
```

fredrik-bakke commented on src/foundation/copartial-functions.lagda.md:
```diff
@@ -0,0 +1,72 @@
+[closed modality](orthogonal-factorization-systems.closed-modalities.md).
+
+A value of a copartial function `f` at `a : A` is said to be
+{{#concept "prohibited" Disambiguation="copartial function"}} if the copartial
```
```suggestion
{{#concept "prohibited" Disambiguation="copartial function" Agda=is-prohibited-copartial-function}} if the copartial
```

fredrik-bakke commented on src/foundation/copartial-functions.lagda.md:
```diff
@@ -0,0 +1,72 @@
+{{#concept "prohibited" Disambiguation="copartial function"}} if the copartial
+element `f a` of `B` is prohibited.
+
+**Note:** The topic of copartial functions is not known in the literature, and
+our formalization on this topic should be considered experimental.
```
Are you sure it is not known at all, for instance under a different name? That seems a little odd to me.

fredrik-bakke replied:
Otherwise, I would phrase this slightly differently

fredrik-bakke commented on src/foundation/total-partial-elements.lagda.md:
```diff
@@ -0,0 +1,33 @@
+A [partial element](foundation.partial-elements.md) `a` of `A` is said to be
+{{#concept "total" Disambiguation="partial element"}} if it is defined. The type
+of total partial elements of `A` is
+[equivalent](foundation-core.equivalences.md) to the type `A`
```
```suggestion
[equivalent](foundation-core.equivalences.md) to the type `A`.
```

fredrik-bakke commented on src/foundation/total-partial-functions.lagda.md:
```diff
@@ -0,0 +1,50 @@
+A [partial function](foundation.partial-functions.md) `f : A → B` is said to be
+{{#concept "total" Disambiguation="partial function"}} if the
+[partial element](foundation.partial-elements.md) `f a` of `B` is defined for
+every `a : A`. The type of total partial functions from `A` to `B` is equivalent
```
```suggestion
every `a : A`. The type of total partial functions from `A` to `B` is equivalent
```
```suggestion
every `a : A`. The type of total partial functions from `A` to `B` is [equivalent](foundation-core.equivalences.md)
```

fredrik-bakke commented on src/foundation/copartial-elements.lagda.md:
````diff
@@ -0,0 +1,102 @@
+A {{#concept "copartial element"}} of a type `A` is an element of type
+
+```text
+  Σ (Q : Prop), A * Q
+```
+
+where the type `P * A` is the
+[join](synthetic-homotopy-theory.joins-of-types.md) of `P` and `A`. We say that
+a copartial element `(P , u)` is
+{{#concept "prohibited" Disambiguation="copartial element"}} if the proposition
+`P` holds.
+
+In order to compare copartial elements with
+[partial elements](foundation.partial-elements.md), note that we have the
+following [pullback squares](foundation.pullback-squares.md)
+
+```text
+  A -----> Σ (Q : Prop), A * Q        1 -----> Σ (P : Prop), (P → A)
+  |                |                  |                |
+  |                |                  |                |
+  V                V                  V                V
+  1 -----------> Prop                 1 -----------> Prop
+          ⊥                                   ⊥
+
+  1 -----> Σ (Q : Prop), A * Q        A -----> Σ (P : Prop), (P → A)
+  |                |                  |                |
+  |                |                  |                |
+  V                V                  V                V
+  1 -----------> Prop                 1 -----------> Prop
+          ⊤                                   ⊤
+```
+
+Note that we make use of the
+[closed modalities](orthogonal-factorization-systems.closed-modalities.md)
+`A ↦ P * A` in the formulation of copartial element, whereas the formulation of
````
This text is inconsistent between using `Q` or `P` for the proposition associated to the open modality. One resolution is:
```suggestion
  Σ (Q : Prop), A * Q
```

where the type `A * Q` is the
[join](synthetic-homotopy-theory.joins-of-types.md) of `Q` and `A`. We say that
a copartial element `(Q , u)` is
{{#concept "prohibited" Disambiguation="copartial element"}} if the proposition
`Q` holds.

In order to compare copartial elements with
[partial elements](foundation.partial-elements.md), note that we have the
following [pullback squares](foundation.pullback-squares.md)

```text
  A -----> Σ (Q : Prop), A * Q        1 -----> Σ (P : Prop), (P → A)
  |                |                  |                |
  |                |                  |                |
  V                V                  V                V
  1 -----------> Prop                 1 -----------> Prop
          ⊥                                   ⊥

  1 -----> Σ (Q : Prop), A * Q        A -----> Σ (P : Prop), (P → A)
  |                |                  |                |
  |                |                  |                |
  V                V                  V                V
  1 -----------> Prop                 1 -----------> Prop
          ⊤                                   ⊤
```

Note that we make use of the
[closed modalities](orthogonal-factorization-systems.closed-modalities.md)
`A ↦ A * Q` in the formulation of copartial element, whereas the formulation of
```

fredrik-bakke commented on src/foundation/copartial-elements.lagda.md:
```diff
@@ -0,0 +1,102 @@
+
+## Idea
+
+A {{#concept "copartial element"}} of a type `A` is an element of type
```
```suggestion
A {{#concept "copartial element" Agda=copartial-element}} of a type `A` is an element of type
```

fredrik-bakke commented on src/foundation/copartial-elements.lagda.md:
```diff
@@ -0,0 +1,102 @@
+where the type `P * A` is the
+[join](synthetic-homotopy-theory.joins-of-types.md) of `P` and `A`. We say that
+a copartial element `(P , u)` is
+{{#concept "prohibited" Disambiguation="copartial element"}} if the proposition
```
```suggestion
{{#concept "prohibited" Disambiguation="copartial element" Agda=is-prohibited-copartial-element}} if the proposition
```

fredrik-bakke reviewed (approved):
Neat! Your idea for copartial elements makes a lot of sense.

VojtechStep commented on src/foundation/copartial-functions.lagda.md:
````diff
@@ -0,0 +1,72 @@
+copartial function is a map
+
+```text
+  A → Σ (Q : Prop), A * Q
````
```suggestion
  A → Σ (Q : Prop), B * Q
```

VojtechStep commented on src/foundation/partial-sequences.lagda.md:
```diff
@@ -0,0 +1,53 @@
+## Idea
+
+A {{#concept "partial sequence"}} of elements of a type `A` is a
+[partial function](foundation.partial-functions.md) from `ℕ` → `A`. In other
```
I've always found this typographical choice a little confusing: a partial sequence is not "a partial function from `ℕ → A`"; it's either "a partial function from `ℕ` to `A`", or "a partial function `ℕ → A`"

fredrik-bakke replied:
Oh, you're right, I didn't notice the strange grammar

EgbertRijke commented:
Instead of calling copartial elements "prohibited" if the underlying proposition holds, another suggestion would be to call them "erased". That would be more clearly dual to "defined". Reviewers, what do you think?

EgbertRijke commented:
Then they could be called "wiped functions" 😁

fredrik-bakke commented:
What role would `P × B` play in this context? After all, this corresponds to the only internal definable coreflective subuniverses

EgbertRijke commented:
Good question! I thought about that option earlier today, but I don't remember why I dismissed it. Maybe I shouldn't have:)

VojtechStep commented:
> another suggestion would be to call them "erased"

I like "erased" 👍

EgbertRijke commented:
I have addressed all the comments. Furthermore, I have also added some informal text explaining how partial functions can be viewed as composites of polynomial endofunctors, and how copartial functions can be viewed as pushout-products. I have not added the formalizations of these insights yet, but it is starting to look very interesting to pursue these remarks in a future pull request.

fredrik-bakke commented on src/foundation/copartial-elements.lagda.md:
````diff
@@ -0,0 +1,141 @@
+
+```text
+  A -----> Σ (Q : Prop), A * Q        1 -----> Σ (P : Prop), (P → A)
+  |                |                  |                |
+  |                |                  |                |
+  V                V                  V                V
+  1 -----------> Prop                 1 -----------> Prop
+          F                                   F
+
+  1 -----> Σ (Q : Prop), A * Q        A -----> Σ (P : Prop), (P → A)
+  |                |                  |                |
````
```suggestion
  | ⌟              |                  | ⌟              |
  |                |                  |                |
  V                V                  V                V
  1 -----------> Prop                 1 -----------> Prop
          F                                   F

  1 -----> Σ (Q : Prop), A * Q        A -----> Σ (P : Prop), (P → A)
  | ⌟              |                  | ⌟              |
```

fredrik-bakke commented on src/foundation/partial-sequences.lagda.md:
````diff
@@ -0,0 +1,53 @@
+words, a partial sequence is a map
+
+```text
+  ℕ → Σ (P : Prop), (P → A)
````
You could also do dependent partial sequences of course :)

EgbertRijke commented on src/foundation/copartial-elements.lagda.md:
````diff
@@ -0,0 +1,141 @@
+                                       P Q ↦ P * Q
+```
+
+By this [morphism of arrows](foundation.morphisms-arrows.md) it follows that
+there is a morphism of arrows
+
+```text
+  join-copartial-element : (A □ T) □ T → A □ T,
+```
+
+i.e., that copartial copartial elements induce copartial elements. These
+considerations allow us to compose
+[copartial functions](foundation.copartial-functions.md).
````
I don't think I'm using monadicity here. Under the hood, associativity of the join is used. If you have something specific in mind, could you add it as a suggestion?

fredrik-bakke reviewed (approved):
Looks good to me!
